src/clipper.cpp

```diff
@@ -466,14 +482,97 @@ std::vector<Compound> islandDetection(Compound& old_compound) {
 //
 //    return results;
 //}
+////////////////////////////////////////////////////////////////////////////////
+
+// pre-fracture for partial fracture
+void preFracture(const Compound& compound, const Pattern& pattern, Compound& inside, Compound& outside) {
+    ///////////////////////////////////////////////////
+    // first create a cube sm for a given pattern
+    Surface_mesh sm; 
+    auto mi = pattern.getMin(); 
+    auto ma = pattern.getMax();
+    // put in all vertices 
+    sm.add_vertex(Point_3(mi.x(), mi.y(), mi.z()));
+    sm.add_vertex(Point_3(mi.x(), mi.y(), ma.z()));
+    sm.add_vertex(Point_3(mi.x(), ma.y(), mi.z()));
+    sm.add_vertex(Point_3(mi.x(), ma.y(), ma.z()));
+    sm.add_vertex(Point_3(ma.x(), mi.y(), mi.z()));
+    sm.add_vertex(Point_3(ma.x(), mi.y(), ma.z()));
+    sm.add_vertex(Point_3(ma.x(), ma.y(), mi.z()));
+    sm.add_vertex(Point_3(ma.x(), ma.y(), ma.z()));
+    // put in all faces 
+    vector<vector<int>> faces{ {0,6,4},{0,2,6},{0,3,2},{0,1,3},
+                               {2,7,6},{2,3,7},{4,6,7},{4,7,5},
+                               {0,4,5},{0,5,1},{1,5,7},{1,7,3} };
+    for (const auto& f : faces) {
+        std::vector<Vertex_descriptor> indices;
+        for (int i : f) {
+            indices.push_back(Vertex_descriptor(i));
+        }
+        sm.add_face(indices);
+    }
+    // triangualte all non-triangle faces
+    PMP::triangulate_faces(sm);
+    ///////////////////////////////////////////////////
+    ABtree tree;
+    tree.add_mesh(sm, PMP::parameters::apply_per_connected_component(false));
+    for (auto& com : compound.convexes) tree.add_mesh(com->convexMesh, PMP::parameters::apply_per_connected_component(false));
+    std::unordered_set<int> intersected;
+    for (auto& intersection : tree.get_all_intersections_and_inclusions(0)) {
+        intersected.insert(intersection.first - 1);
+    }
+    for (int i = 0; i < compound.convexes.size(); ++i) {
```

Comment:
Do we really need this `for-loop`? Can we just use `intersected` to approximate `inside` and others left in compound default to `outside`, since the main fracturing/clipping happens later down the pipeline? This way we can save some computational resources.

Reply:
这个真用不了多少resource，你用一个outside那你还要remove inside的convex，unorder_set无所谓的